### PR TITLE
LibWeb: Add missing property and methods for history object

### PR DESCRIPTION
**Pull request by leeight**
We provide `length` property and `go` / `back` / `forward` methods implementation here.

### CODE REVIEW

File: Userland/Libraries/LibWeb/HTML/History.cpp [hunk 1]
```diff
@@ -44,6 +44,52 @@ WebIDL::ExceptionOr<void> History::replace_state(JS::Value data, String const&,
     return shared_history_push_replace_state(data, url, IsPush::No);
 }
 
+WebIDL::ExceptionOr<u64> History::length() const
+{
+    // 1. Let document be this's associated Document.
+
+    // 2. If document is not fully active, then throw a "SecurityError" DOMException.
+    if (!m_associated_document->is_fully_active())
+        return WebIDL::SecurityError::create(realm(), "Cannot perform length on a document that isn't fully active.");
```
Comment (linusg) [inline Userland/Libraries/LibWeb/HTML/History.cpp:53]:
```suggestion
        return WebIDL::SecurityError::create(realm(), "Cannot perform length on a document that isn't fully active."sv);
```

File: Userland/Libraries/LibWeb/HTML/History.cpp [hunk 2]
```diff
@@ -44,6 +44,52 @@ WebIDL::ExceptionOr<void> History::replace_state(JS::Value data, String const&,
...
+
+    auto const* browsing_context = m_associated_document->browsing_context();
+    if (!browsing_context)
+        return WebIDL::SecurityError::create(realm(), "Cannot perform length on a document that isn't fully active.");
```
Comment (linusg) [inline Userland/Libraries/LibWeb/HTML/History.cpp:57]: When would this be the case after the previous check?

File: Userland/Libraries/LibWeb/HTML/History.cpp [hunk 3]
```diff
@@ -44,6 +44,52 @@ WebIDL::ExceptionOr<void> History::replace_state(JS::Value data, String const&,
     return shared_history_push_replace_state(data, url, IsPush::No);
 }
 
+WebIDL::ExceptionOr<u64> History::length() const
```
Comment (linusg) [inline Userland/Libraries/LibWeb/HTML/History.cpp:47]: Missing spec URL

File: Userland/Libraries/LibWeb/HTML/History.cpp [hunk 4]
```diff
@@ -44,6 +44,52 @@ WebIDL::ExceptionOr<void> History::replace_state(JS::Value data, String const&,
...
+{
+    // 1. Let document be this's associated Document.
```
Comment (linusg) [inline Userland/Libraries/LibWeb/HTML/History.cpp:49]: Incorrect comments here, these are not from https://html.spec.whatwg.org/multipage/history.html#dom-history-length

File: Userland/Libraries/LibWeb/HTML/History.cpp [hunk 5]
```diff
@@ -44,6 +44,52 @@ WebIDL::ExceptionOr<void> History::replace_state(JS::Value data, String const&,
...
+History::go(long delta = 0)
+{
+    // 1. Let document be this's associated Document.
+
+    // 2. If document is not fully active, then throw a "SecurityError" DOMException.
+    if (!m_associated_document->is_fully_active())
+        return WebIDL::SecurityError::create(realm(), "Cannot perform go on a document that isn't fully active.");
+    auto* browsing_context = m_associated_document->browsing_context();
+    if (!browsing_context)
+        return WebIDL::SecurityError::create(realm(), "Cannot perform go on a document that isn't fully active.");
```
Comment (linusg) [inline Userland/Libraries/LibWeb/HTML/History.cpp:73]: Same as above

File: Userland/Libraries/LibWeb/HTML/History.cpp [hunk 6]
```diff
@@ -44,6 +44,52 @@ WebIDL::ExceptionOr<void> History::replace_state(JS::Value data, String const&,
...
+    // 3. If delta is 0, then act as if the location.reload() method was called, and return.
+    // 4. Traverse the history by a delta with delta and document's browsing context.
+    auto entry_index = browsing_context->session_history_index();
+    return browsing_context->traverse_the_history(entry_index + delta, HistoryHandlingBehavior::Reload);
+}
+
+// https://html.spec.whatwg.org/multipage/history.html#dom-history-back
+WebIDL::ExceptionOr<void> History::back()
+{
+    return go(-1);
```
Comment (linusg) [inline Userland/Libraries/LibWeb/HTML/History.cpp:84]: Please still add the comments here

File: Userland/Libraries/LibWeb/HTML/History.cpp [hunk 7]
```diff
@@ -44,6 +44,52 @@ WebIDL::ExceptionOr<void> History::replace_state(JS::Value data, String const&,
...
+}
+
+// https://html.spec.whatwg.org/multipage/history.html#dom-history-forward
+WebIDL::ExceptionOr<void> History::forward()
+{
+    return go(1);
```
Comment (linusg) [inline Userland/Libraries/LibWeb/HTML/History.cpp:90]: Likewise

File: Userland/Libraries/LibWeb/HTML/History.idl [hunk 8]
```diff
@@ -1,5 +1,12 @@
+// https://html.spec.whatwg.org/multipage/history.html#the-history-interface
 [Exposed=Window]
 interface History {
+    readonly attribute unsigned long length;
```
Comment (linusg) [inline Userland/Libraries/LibWeb/HTML/History.idl:4]: This is not mentioned in the commit message, consider splitting this PR up into multiple commits

File: Userland/Libraries/LibWeb/HTML/History.idl [hunk 9]
```diff
@@ -1,5 +1,12 @@
...
+    // attribute ScrollRestoration scrollRestoration;
+    // readonly attribute any state;
```
Comment (linusg) [inline Userland/Libraries/LibWeb/HTML/History.idl:6]:
```suggestion
    // FIXME: attribute ScrollRestoration scrollRestoration;
    // FIXME: readonly attribute any state;
```

**Review by linusg — CHANGES_REQUESTED** [review]

File: Userland/Libraries/LibWeb/HTML/History.cpp [hunk 10]
```diff
@@ -44,14 +44,68 @@ WebIDL::ExceptionOr<void> History::replace_state(JS::Value data, String const&,
     return shared_history_push_replace_state(data, url, IsPush::No);
 }
 
+// https://html.spec.whatwg.org/multipage/history.html#dom-history-length
+WebIDL::ExceptionOr<u64> History::length() const
+{
+    // 1. Let document be history's associated Document. (NOTE: Not necessary)
```
Comment (linusg) [inline Userland/Libraries/LibWeb/HTML/History.cpp:50]: Still wrong, the steps from https://html.spec.whatwg.org/multipage/history.html#dom-history-length are:

```
1. If this's associated Document is not fully active, then throw a "SecurityError" DOMException.
2. Return the number of entries in the top-level browsing context's joint session history.
```

File: Userland/Libraries/LibWeb/HTML/History.cpp [hunk 11]
```diff
@@ -44,14 +44,68 @@ WebIDL::ExceptionOr<void> History::replace_state(JS::Value data, String const&,
...
+    if (!m_associated_document->is_fully_active())
+        return WebIDL::SecurityError::create(realm(), "Cannot perform length on a document that isn't fully active."sv);
+
+    auto const* browsing_context = m_associated_document->browsing_context();
+    return browsing_context->session_history().size();
+}
+
+// https://html.spec.whatwg.org/multipage/history.html#dom-history-go
+WebIDL::ExceptionOr<void>
+History::go(long delta = 0)
```
Comment (linusg) [inline Userland/Libraries/LibWeb/HTML/History.cpp:62]: Broken formatting here

File: Userland/Libraries/LibWeb/HTML/History.cpp (hunk 10 above)
Comment (leeight) [reply Userland/Libraries/LibWeb/HTML/History.cpp:50]: OK

File: Userland/Libraries/LibWeb/HTML/History.cpp [hunk 12]
```diff
@@ -44,14 +44,71 @@ WebIDL::ExceptionOr<void> History::replace_state(JS::Value data, String const&,
...
+
+    // 3. If delta is 0, then act as if the location.reload() method was called, and return.
+    // 4. Traverse the history by a delta with delta and document's browsing context.
+    auto* browsing_context = m_associated_document->browsing_context();
+    auto current_entry_index = browsing_context->session_history_index();
+    auto next_entry_index = current_entry_index + delta;
+    auto sessions = browsing_context->session_history();
+    if (next_entry_index < sessions.size()) {
+        auto next_entry = sessions.at(next_entry_index);
+        browsing_context->loader().load(next_entry.url, FrameLoader::Type::Reload);
```
Comment (linusg) [inline Userland/Libraries/LibWeb/HTML/History.cpp:76]: This is far from spec compliant, it's fine as a first basic version but please mark the comment `// FIXME: 4. Traverse ...` so we can find this later.

File: Userland/Libraries/LibWeb/HTML/History.cpp [hunk 13]
```diff
@@ -44,14 +44,71 @@ WebIDL::ExceptionOr<void> History::replace_state(JS::Value data, String const&,
...
+    // 2. If document is not fully active, then throw a "SecurityError" DOMException.
+    if (!m_associated_document->is_fully_active())
+        return WebIDL::SecurityError::create(realm(), "Cannot perform go on a document that isn't fully active."sv);
+
+    // 3. If delta is 0, then act as if the location.reload() method was called, and return.
+    // 4. Traverse the history by a delta with delta and document's browsing context.
+    auto* browsing_context = m_associated_document->browsing_context();
+    auto current_entry_index = browsing_context->session_history_index();
+    auto next_entry_index = current_entry_index + delta;
+    auto sessions = browsing_context->session_history();
```
Comment (linusg) [inline Userland/Libraries/LibWeb/HTML/History.cpp:73]:
```suggestion
    auto const& sessions = browsing_context->session_history();
```

No need to make a copy

File: Userland/Libraries/LibWeb/HTML/History.cpp [hunk 14]
```diff
@@ -44,14 +44,71 @@ WebIDL::ExceptionOr<void> History::replace_state(JS::Value data, String const&,
...
+    if (next_entry_index < sessions.size()) {
+        auto next_entry = sessions.at(next_entry_index);
```
Comment (linusg) [inline Userland/Libraries/LibWeb/HTML/History.cpp:75]:
```suggestion
    auto const& next_entry = sessions.at(next_entry_index);
```

No need to make a copy

File: Userland/Libraries/LibWeb/HTML/History.cpp [hunk 15]
```diff
@@ -44,14 +44,71 @@ WebIDL::ExceptionOr<void> History::replace_state(JS::Value data, String const&,
...
+// https://html.spec.whatwg.org/multipage/history.html#dom-history-length
+WebIDL::ExceptionOr<u64> History::length() const
+{
+    // 1. If this's associated Document is not fully active, then throw a "SecurityError" DOMException.
+    if (!m_associated_document->is_fully_active())
+        return WebIDL::SecurityError::create(realm(), "Cannot perform length on a document that isn't fully active."sv);
+
+    // 2. Return the number of entries in the top-level browsing context's joint session history.
+    auto const* browsing_context = m_associated_document->browsing_context();
+    return browsing_context->session_history().size();
```
Comment (ADKaster) [inline Userland/Libraries/LibWeb/HTML/History.cpp:56]: If we're don't have the concept of "joint session history" https://html.spec.whatwg.org/multipage/history.html#joint-session-history implemented in LibWeb yet, then this should have a FIXME to use the joint session history.

> The joint session history of a [top-level browsing context](https://html.spec.whatwg.org/multipage/browsers.html#top-level-browsing-context) is the union of all the [session histories](https://html.spec.whatwg.org/multipage/history.html#session-history) of all [browsing contexts](https://html.spec.whatwg.org/multipage/browsers.html#concept-document-bc) of all the [fully active](https://html.spec.whatwg.org/multipage/browsers.html#fully-active) [Document](https://html.spec.whatwg.org/multipage/dom.html#document) objects that share that [top-level browsing context](https://html.spec.whatwg.org/multipage/browsers.html#top-level-browsing-context), with all the entries that are [current entries](https://html.spec.whatwg.org/multipage/history.html#current-entry) in their respective [session histories](https://html.spec.whatwg.org/multipage/history.html#session-history) removed except for the [current entry of the joint session history](https://html.spec.whatwg.org/multipage/history.html#current-entry-of-the-joint-session-history).

Comment (leeight) [reply Userland/Libraries/LibWeb/HTML/History.cpp:56]: ok

File: Userland/Libraries/LibWeb/HTML/History.cpp [hunk 16]
```diff
@@ -44,14 +44,74 @@ WebIDL::ExceptionOr<void> History::replace_state(JS::Value data, String const&,
...
+{
+    // 1. If this's associated Document is not fully active, then throw a "SecurityError" DOMException.
+    if (!m_associated_document->is_fully_active())
+        return WebIDL::SecurityError::create(realm(), "Cannot perform length on a document that isn't fully active."sv);
+
+    // 2. Return the number of entries in the top-level browsing context's joint session history.
+    auto const* browsing_context = m_associated_document->browsing_context();
+
+    // FIXME: We're don't have the concept of "joint session history", this is ad-hoc implementation.
+    //        See more https://html.spec.whatwg.org/multipage/history.html#joint-session-history
```
Comment (linusg) [inline Userland/Libraries/LibWeb/HTML/History.cpp:58]:
```suggestion
    // FIXME: We don't have the concept of "joint session history", this is an ad-hoc implementation.
    //        See: https://html.spec.whatwg.org/multipage/history.html#joint-session-history
```

Comment (leeight) [reply Userland/Libraries/LibWeb/HTML/History.cpp]: DONE

**Pull request merged**